sm_value_store/include/sm/value_store/LayeredValueStore.hpp

```diff
@@ -14,10 +14,10 @@ class LayeredValueStore : public ValueStore {
   LayeredValueStore(ValueStores && ... vss) {
     add(vss...);
   }
-  template <typename ValueStoreA, typename ValueStoreB, typename ...ValueStores>
-  void add(ValueStoreA && vsA, ValueStoreB && vsB, ValueStores && ... vss) {
+  template <typename ValueStoreA, typename ...ValueStores>
+  void add(ValueStoreA && vsA, ValueStores && ... vss) {
     add(vsA);
-    add(vsB, vss...);
```

Comment:
This looks dangerous to me. It makes the add with a single argument potentially ambiguous or even worse, favor this template function leading to a infinite recursion, no? What was the reason for this change?

Reply:
It did not compile before, so I just assumed that it was not used and thus not compiled (since template):
```
/home/yannick/catkin_ws/src/schweizer_messer/sm_value_store/include/sm/value_store/LayeredValueStore.hpp:15:5: error: no matching function for call to ‘sm::value_store::LayeredValueStore::add(const sm::PropertyTree&)’
   15 |     add(vss...);
   ```
Infinite recursion can happen because variadic templates can be empty right?
I'm pretty sure I use something similar in quite some places

Reply:
Yes, the worry is about an empty variadric argument.
This can work well, if you  are sure that the function with a single argument always takes precedence over the variadric template, which it may often do, e.g. because it is not a template. But from memory it was a risk here (so I assume the single argument function is also a template).

Reply:
No, it isn't, but I think there is a case even for a non-template to match worse: by requiring a (user) conversion. Thus, with your change I think we'll get the unbounded recursion, when having an argument that doesn't have decltype std::shared_ptr<ValueStore> or ValueStoreRef & p.. That could still be a risk for the other cases where you are using the simple version. I wonder how this is properly dealt with..